contracts/LogarithmReverseDutchAuction.sol

```diff
@@ -0,0 +1,228 @@
+        address biddingToken,
+        uint256 startingPrice,
+        uint256 reservedPrice,
+        uint256 decayFactor,
+        uint256 duration
+    ) external{
+        require(bytes(name).length>0,"Name must be present");
+        require(reservedPrice>=0,"Reserved price cannot be negative");
+        require(startingPrice>=0,"Starting price cannot be negative");
+        require(startingPrice>=reservedPrice,"Starting price should be higher than reserved price");
+        require(duration>0,"Duration must be greater than zero seconds");
+        require(decayFactor>=0,"Decay factor cannot be negative");
+        require(auctionedToken!=address(0),"Auctioned token cannot be zero address");
+        require(biddingToken!=address(0),"Bidding token cannot be zero address");
+        //decay Factor is scaled with 10^3 to ensure precision upto three decimal points
+
+        if(auctionType == AuctionType.NFT){
+            require(IERC721(auctionedToken).ownerOf(auctionedTokenIdOrAmount)==msg.sender,"Caller must be the owner");
+            IERC721(auctionedToken).safeTransferFrom(msg.sender,address(this),auctionedTokenIdOrAmount);
+        }else{
+            require(IERC20(auctionedToken).balanceOf(msg.sender)>=auctionedTokenIdOrAmount,"Insufficient balance");
+            SafeERC20.safeTransferFrom(IERC20(auctionedToken),msg.sender,address(this),auctionedTokenIdOrAmount);
+        }
+
+        uint256 deadline=block.timestamp+duration;
+        uint256 scalingFactor=log2Fixed(1+(decayFactor*duration)/1e3,6);//log2(1+k*duration/1000) with 6 decimal precision
+        require(scalingFactor > 0, "Scaling factor must be greater than zero");
+        auctions[auctionCounter] = AuctionData({
+            id: auctionCounter,
+            name:name,
+            description:description,
+            imgUrl:imgUrl,
+            auctioneer: msg.sender,
+            auctionType: auctionType,
+            auctionedToken: auctionedToken,
+            auctionedTokenIdOrAmount: auctionedTokenIdOrAmount,
+            biddingToken: biddingToken,
+            startingPrice: startingPrice,
+            availableFunds: 0,
+            reservedPrice: reservedPrice,
+            decayFactor: decayFactor,
+            winner: msg.sender,
+            deadline: deadline,
+            duration: duration,
+            scalingFactor: scalingFactor,
+            isClaimed: false
+        });
+
+        emit AuctionCreated(
+            auctionCounter++,
+            name,
+            description,
+            imgUrl,
+            msg.sender,
+            auctionType,
+            auctionedToken,
+            auctionedTokenIdOrAmount,
+            biddingToken,
+            startingPrice,
+            reservedPrice,
+            decayFactor,
+            deadline
+        );
+    }
+
+    //Returns the logarithm base 2 of x scaled by 10^18 with a specified number of fractional bits.
+    function log2Fixed(uint256 x,uint8 fracBits) internal pure returns (uint256){
```

Comment:
Isn't there a library that we could use, instead of implementing our own logarithm functions?